### ISSUE TITLE
This plugin is not compatible with HB 2.0

### ISSUE THREAD

**Issue by pfrasun** [Jason-Morcos/homebridge-magic-occupancy#50]
This plugin is sadly not compatible with HB 2.0. After installing HB 2.0 Beta 11 plugin does not start properly and ends in the endless restart loops. Below errors show in the log:
```
[19/08/2024, 12:33:28] [homebridge-magic-occupancy] Restarting Process...
[19/08/2024, 12:33:29] [homebridge-magic-occupancy] Launched child bridge with PID 531864
[19/08/2024, 12:33:29] Registering accessory 'homebridge-magic-occupancy.MagicOccupancy'
[19/08/2024, 12:33:30] [homebridge-magic-occupancy] Loaded homebridge-magic-occupancy v3.3.0 child bridge successfully with 5 accessories
HAP-NodeJS WARNING: The accessory 'homebridge-magic-occupancy' has an invalid 'Name' characteristic ('homebridge-magic-occupancy'). Please use only alphanumeric, space, and apostrophe characters. Ensure it starts and ends with an alphabetic or numeric character, and avoid emojis. This may prevent the accessory from being added in the Home App or cause unresponsiveness.
[19/08/2024, 12:33:30] Loaded 0 cached accessories from cachedAccessories.0EDD6B87BFA4.

/var/lib/homebridge/node_modules/homebridge-magic-occupancy/index.js:43
        Characteristic.call(
                       ^
TypeError: Class constructor Characteristic cannot be invoked without 'new'
    at new Characteristic.TimeoutDelay (/var/lib/homebridge/node_modules/homebridge-magic-occupancy/index.js:43:24)
    at OccupancySensor.addCharacteristic (/var/lib/homebridge/node_modules/homebridge/node_modules/hap-nodejs/src/lib/Service.ts:518:57)
    at new MagicOccupancy (/var/lib/homebridge/node_modules/homebridge-magic-occupancy/index.js:158:31)
    at ChildBridgeFork.startBridge (/var/lib/homebridge/node_modules/homebridge/src/childBridgeFork.ts:177:52)
[19/08/2024, 12:33:30] [homebridge-magic-occupancy] Child bridge process ended
[19/08/2024, 12:33:30] [homebridge-magic-occupancy] Process Ended. Code: 1, Signal: null
[19/08/2024, 12:33:37] [homebridge-magic-occupancy] Restarting Process...
[19/08/2024, 12:33:39] [homebridge-magic-occupancy] Launched child bridge with PID 531882
[19/08/2024, 12:33:39] Registering accessory 'homebridge-magic-occupancy.MagicOccupancy'
[19/08/2024, 12:33:39] [homebridge-magic-occupancy] Loaded homebridge-magic-occupancy v3.3.0 child bridge successfully with 5 accessories
HAP-NodeJS WARNING: The accessory 'homebridge-magic-occupancy' has an invalid 'Name' characteristic ('homebridge-magic-occupancy'). Please use only alphanumeric, space, and apostrophe characters. Ensure it starts and ends with an alphabetic or numeric character, and avoid emojis. This may prevent the accessory from being added in the Home App or cause unresponsiveness.
[19/08/2024, 12:33:39] Loaded 0 cached accessories from cachedAccessories.0EDD6B87BFA4.

/var/lib/homebridge/node_modules/homebridge-magic-occupancy/index.js:43
        Characteristic.call(
                       ^
TypeError: Class constructor Characteristic cannot be invoked without 'new'
    at new Characteristic.TimeoutDelay (/var/lib/homebridge/node_modules/homebridge-magic-occupancy/index.js:43:24)
    at OccupancySensor.addCharacteristic (/var/lib/homebridge/node_modules/homebridge/node_modules/hap-nodejs/src/lib/Service.ts:518:57)
    at new MagicOccupancy (/var/lib/homebridge/node_modules/homebridge-magic-occupancy/index.js:158:31)
    at ChildBridgeFork.startBridge (/var/lib/homebridge/node_modules/homebridge/src/childBridgeFork.ts:177:52)
[19/08/2024, 12:33:39] [homebridge-magic-occupancy] Child bridge process ended
[19/08/2024, 12:33:39] [homebridge-magic-occupancy] Process Ended. Code: 1, Signal: null
```

**Comment by Jason-Morcos** [comment]
If someone wants to take a stab opening a PR that adds v2 support, I'd be glad to review! Otherwise I'll take a look in a couple weeks when I have some bandwidth

**Comment by pfrasun** [comment]
Just as a reminder. Do you plan to do something about it?

**Comment by Jason-Morcos** [comment]
Thanks for the reminder! Turned out to require fairly substantial changes but the v4 version should support both Homebridge 1 and 2

**Comment by pfrasun** [comment]
Version 4 does work with HB2, but the plugin itself stopped working for me. Switches do not trigger occupancy sensor.

**Comment by Jason-Morcos** [comment]
Odd - it appears to be working for me
@pfrasun are you running v1 or v2 of homebridge?
Anything showing up in the logs that might be useful for debugging?
Is it just specific types of switches that aren't working? Can you share the problematic config?

**Comment by pfrasun** [comment]
That is really strange. I tried in both configs:

- 1.8.4 - plug-in worked, but stateful switches were not turning off presence sensor, only working one was master reset
- 2.0.0. Beta 22 - behaviour was exactly same, but additionally plugin crashed and was showing errors (related to the name of the bridge)

There are no errors in logs related to the issue. Switch events are not shown it means the switch does not work. New installation did not solve it.

**Comment by Jason-Morcos** [comment]
Stateful switches shouldn't turn off the presence sensor unless the stay occupied delay is set to zero (otherwise they'll stay on after the switch flips off for the stay occupied delay duration).
Have you tried reverting to 3.x to ensure it's behaving differently in v4 than it did in v3? If it's definitely borked, can you share the part of your config for the plugin so I can try to replicate your exact setup?

**Comment by pfrasun** [comment]
Yes, 3.3.0 works ans intended.
I use it with presence sensors having no delay, but also sensors with delay don’t work, I tried it.

**Comment by Jason-Morcos** [comment]
I think I figured it out and fully fixed in 4.0.1! I removed a previously-undocumented private service (basically home bridge HAP changes appear to make these kind of dynamic characteristics not as kosher anymore). Let me know if this solves it
https://github.com/Jason-Morcos/homebridge-magic-occupancy/pull/52
